### DISCUSSION TITLE
Parsing json with unnamed array

### DISCUSSION THREAD

**Discussion by ankur90Git** [microsoft/PSRule#1131] [Q&A] [answered]
I am trying to do PSRule validation on below json file, which is having an unnamed array with 3 elements. I am looking to fetch the value of field 'RoleDefinitionName' of the first element only and validate its value. 

I tried using this expression but its returning me value from all three child elements. How can we handle this unnamed array for fetching a particular value. Any help or suggestion would be highly appreciated. Thankyou.

$ActualValue = $TargetObject.RoleDefinitionName

![image](https://user-images.githubusercontent.com/84568209/175067066-bd924772-b944-4ac1-8b09-0a14344a3921.png)

![image](https://user-images.githubusercontent.com/84568209/175066944-956a8c20-32fa-409c-b205-9872800b7b93.png)

[
  {
    "RoleAssignmentName": "d57dab50-20b5-491f-aed6-489b738c411c",
    "RoleAssignmentId": "fdjkhwiuofhwdnflo",
    "Scope": "foihweohfopihpon",
    "DisplayName": "EUWUEFDWEHF",
    "SignInName": null,
    "RoleDefinitionName": "App Configuration Data Reader",
    "RoleDefinitionId": "qf8727839-63e1-4d78-a4de-23423daf",
    "ObjectId": "wft3798-0191-1111-3333-4444",
    "ObjectType": "ServicePrincipal",
    "CanDelegate": false,
    "Description": null,
    "ConditionVersion": null,
    "Condition": null
  },
  {
    "RoleAssignmentName": "d57dab50-20b5-491f-aed6-489b738c411c",
    "RoleAssignmentId": "fdjkhwiuofhwdnflo",
    "Scope": "foihweohfopihpon",
    "DisplayName": "EUWUEFDWEHF",
    "SignInName": null,
    "RoleDefinitionName": "Azure Service Bus Data Receiver",
    "RoleDefinitionId": "qf8727839-63e1-4d78-a4de-23423daf",
    "ObjectId": "wft3798-0191-1111-3333-4444",
    "ObjectType": "ServicePrincipal",,
    "CanDelegate": false,
    "Description": null,
    "ConditionVersion": null,
    "Condition": null
  },
  {
    "RoleAssignmentName": "d57dab50-20b5-491f-aed6-489b738c411c",
    "RoleAssignmentId": "fdjkhwiuofhwdnflo",
    "Scope": "foihweohfopihpon",
    "DisplayName": "EUWUEFDWEHF",
    "SignInName": null,
    "RoleDefinitionName": "Azure Service Bus Data Receiver",
    "RoleDefinitionId": "qf8727839-63e1-4d78-a4de-23423daf",
    "ObjectId": "wft3798-0191-1111-3333-4444",
    "ObjectType": "ServicePrincipal",
    "CanDelegate": false,
    "Description": null,
    "ConditionVersion": null,
    "Condition": null
  }
]

**Comment by BernieWhite** [comment, marked as answer]
@ankur90Git By default rules run against all input objects, however you can use conditions to limit which objects that apply to by configuring a pre-condition. When set, rules will only apply to objects that return `$true` for the pre-condition.

For a single rule, the easiest way would be to use a script pre-condition, like this:

```powershell
# Synopsis: An example script rule with pre-conditions.
Rule 'AppConfigurationPermission' -If { $TargetObject.RoleDefinitionName -eq 'App Configuration Data Reader' } {
    # Rule condition
}
```

While the other objects would still be discovered, the rule will only run against the first object. Which is probably what you want because there may be other rules for the other objects. You can disable the warning generated for objects not matching any rules by setting [Execution.NotProcessedWarning](https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Options/#executioninconclusivewarning).

There are some more detail here: [Using pre-conditions](https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Rules/#using-pre-conditions)

---

I hope that helps.

**Comment by ankur90Git** [comment]
Thankyou so much for answering my question. Your answer helped me in solving my problem. 
However, I was unclear about your statement "By default rules run against all input objects". In the above example, its only one JSON input file which is having an array with 3 elements. Then why PSRule is considering it as 3 input objects and running the Rule thrice? Please correct me if I misunderstood any concept here.

**Reply by BernieWhite** [reply]
@ankur90Git No worries. When PSRule is the run with the `-InputPath` parameter the default format used is `Detect` which auto detects files based on their extension and deserializes objects contained within the files based on their format.

See [Input.Format](https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Options/#inputformat).

This means that if the `json` file contains an array at the root of the file, each element of the array would be treated as a separate object, three in your case.

**Reply by ankur90Git** [reply]
@BernieWhite ok thanks. One last question, for such json(having array at the root) any option which we can use so that PS Rule considers it as 1 input only and not 3? Because I have all the input json of this format only and creating rule conditions for such input is getting bit complex. Appreciate your response.

**Reply by BernieWhite** [reply]
@ankur90Git There isn't an easy way to do this by default, but it could be done with a convention. Although maybe binding, selectors, or helper functions make address the concerns.

- [Use custom binding](https://microsoft.github.io/PSRule/v2/scenarios/kubernetes-resources/kubernetes-resources/#use-custom-binding)
- [Selectors](https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Selectors/)
- [Define helper functions](https://microsoft.github.io/PSRule/v2/scenarios/azure-resources/azure-resources/#define-helper-functions)

Writing a convention, you can add this is a separate `.Rule.ps1` file or the same existing one.

```powershell
# Synopsis: Create .JSON file and import whole array object.
Export-PSRuleConvention 'ExportArray' -If { $TargetObject -is [PSRule.Data.InputFileInfo] -and $TargetObject.Extension -eq '.json' } -Begin {
    $json = Get-Content -Path $TargetObject.FullName -Raw | ConvertTo-Json;
    $PSRule.Import($json);
}
```

You would then reference it from you command-line.

```powershell
 Invoke-Psrule -InputPath .\example.json -Path .\.ps-rule\ -Format File -Convention ExportArray
```

Each file would be found then imported.

---

Outside of that maybe you can explain your use case in feature request if none of these options work.

Thanks.
